dask/bag/core.py

```diff
@@ -68,12 +69,34 @@ def lazify(dsk):
     return valmap(lazify_task, dsk)
 
 
+def inline_singleton_lists(dsk):
```

Comment:
`Fuse` only handles linear chains, this is to specifically include nodes with multiple inputs.

Usually we don't want to fuse operations with multiple inputs because those inputs could previously have happened in parallel.  In the case of dask.bag we're willing to destroy potential parallelism if it helps us to avoid reifying an iterator into a list.  We sacrifice potential parallelism for better streaming behavior.

Rolling this into inline_functions sounds like a reasonable idea.